### PR TITLE
Feature/oe 148 new customer field changes

### PR DESCRIPTION
**Pull request by Ron-Junior**
## Goal

As a user, I want some fields to be auto-populated when I go to insert a new customer and make them easier to search through

## Changeset

Acceptance Criteria

- When user goes to insert a new customer, the sales rep dropdown should automatically get populated with that user’s name (weather the user has the role of a sales rep or not) IF the user has an office

- Sales rep dropdown should be a list of people under you (and include yourself IF the user has an office). If I’m just a sales rep, the dropdown will only have myself in it. If I’m an office manager, it will have everyone in my office but it will get pre-populated with my name

- Make so the first option in “setter” is “Self Gen” and have it be the default when any user goes to add a customer

- When “setter” = “Self Gen”, clear the setter commission field and disable it

- Setter should have a list of everyone in the department

- List options in setter and sales rep in alphabetical order (except for the very first option in setter)

- Make both sales rep and setter fields searchable and always show up to 5 results for search

## Tests

- [ ] Go to one energy
- [ ] On dashboard click on add customer
- [ ] Verify if all changes its all right
- [ ] go to dashboard click on created customer
- [ ] verify if its all right

## Review

For the submitter, initial self-review:

- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency between the changeset and the goal stated above
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Complexity - is the proposed change introducing any unjustified complexity? 
- [ ] Thoroughness of added tests and any missing edge cases

### CODE REVIEW

File: app/Models/User.php [hunk 1]
```diff
@@ -124,6 +138,33 @@ public function revokeMastersAccess()
         $this->save();
     }
 
+    public function getPermittedUsers()
+    {
+        if($this->role == 'Admin' || $this->role == 'Owner') {
+            // dd(User::whereHas('office')->get()->sortBy('first_name'));
+            return User::whereHas('office')->get()->sortBy('first_name');
```
Comment (iammayron) [inline app/Models/User.php:145]: You could orderBy on eloquent side before the get() function.

File: app/Models/User.php [hunk 2]
```diff
@@ -124,6 +138,33 @@ public function revokeMastersAccess()
...
+        }
+
+        if($this->role == 'Department Manager') {
+            return User::whereDepartmentId($this->department_id)
+                ->orWhere('users.id', $this->id)
+                ->orWhere('role', 'Region Manager')
+                ->orWhere('role', 'Office Manager')
+                ->orWhere('role', 'Sales Rep')
+                ->orWhere('role', 'Setter')->get()->sortBy('first_name');
```
Comment (iammayron) [inline app/Models/User.php:154]: You could orderBy on eloquent side before the get() function.

File: app/Http/Livewire/Customer/Edit.php [hunk 3]
```diff
@@ -58,6 +65,8 @@ public function render()
     {
         $this->customer->calcComission();
         $this->grossRepComission = $this->calculateGrossRepComission($this->customer);
+        $this->salesReps = user()->getPermittedUsers()->toArray();
+        $this->setters = User::whereDepartmentId($this->departmentId)->orderBy('first_name')->get()->toArray();
```
Comment (iammayron) [inline app/Http/Livewire/Customer/Edit.php:69]: I think that you don't need that toArray() function after get().

File: app/Http/Livewire/Customer/Create.php [hunk 4]
```diff
@@ -50,20 +56,25 @@ class Create extends Component
...
+        $this->setSelfGen();
     }
 
     public function render()
     {
         $this->customer->calcComission();
         $this->customer->calcMargin();
         $this->grossRepComission = $this->calculateGrossRepComission($this->customer);
+        $this->salesReps = user()->getPermittedUsers()->toArray();
+        $this->setters = User::whereDepartmentId($this->departmentId)->orderBy('first_name')->get()->toArray();
```
Comment (iammayron) [inline app/Http/Livewire/Customer/Create.php:77]: I think that you don't need that toArray() function after get().

File: app/Models/User.php [hunk 5]
```diff
@@ -75,6 +75,20 @@ public function managedRegions()
         return $this->hasMany(Region::class, 'region_manager_id');
     }
 
+    public function usersOnManagedRegions()
+    {
+        $offices = $this->hasManyThrough(Office::class, Region::class, 'region_manager_id', 'region_id')->with('users')->get();
+        $users = $offices->reduce(function($users, Office $office) {
+            return $users->mergeRecursive($office->users);
+        }, $users = collect([]))->unique('id');
+        return $users->sortBy('first_name');
```
Comment (iammayron) [inline app/Models/User.php:84]: Relation methods must not have logic that doesn't correspond to the relationship, I think that this reduce function and sortBy function are something particular to a single logic.

**Review by iammayron — CHANGES_REQUESTED** [review]

File: app/Http/Livewire/Customer/Edit.php (hunk 3 above)
Comment (Ron-Junior) [reply app/Http/Livewire/Customer/Edit.php:69]: For some reason, the livewire can not list the collection, and because of this we need toArray

File: app/Models/User.php (hunk 2 above)
Comment (Ron-Junior) [reply app/Models/User.php:154]: For some reason, the livewire can not list the collection, and because of this we need toArray

File: app/Http/Livewire/Customer/Create.php (hunk 4 above)
Comment (Ron-Junior) [reply app/Http/Livewire/Customer/Create.php:77]: For some reason, the livewire can not list the collection, and because of this we need toArray

File: app/Http/Livewire/Customer/Edit.php (hunk 3 above)
Comment (iammayron) [reply app/Http/Livewire/Customer/Edit.php:69]: image

The problem is that you declared the property as an array.

File: app/Http/Livewire/Customer/Create.php (hunk 4 above)
Comment (iammayron) [reply app/Http/Livewire/Customer/Create.php:77]: image

The problem is that you declared the property as an array.

**Review by iammayron — APPROVED** [review]
